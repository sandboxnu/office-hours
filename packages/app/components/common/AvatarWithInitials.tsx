import { Avatar } from "antd";
import React, { ReactElement, Ref } from "react";
import nameToRGB from "../../utils/ColorUtils";
import getInitialsFromName from "../../utils/NameUtils";

type AvatarWithInitalsProps = {
  name: string;
  fontSize?: number;
  style?: any;
  [x: string]: any;
};

export const AvatarWithInitals = React.forwardRef(
  (
    { name, fontSize, style, ...props }: AvatarWithInitalsProps,
    ref: Ref<HTMLElement>
  ): ReactElement => {
    return (
      <Avatar
        {...props}
        ref={ref}
        style={{
          backgroundColor: name ? nameToRGB(name) : "#1abc9c",
          fontSize,
          ...style,
        }}
      >
        {getInitialsFromName(name)}
      </Avatar>
    );
  }
);

AvatarWithInitals.displayName = "AvatarWithInitials";
