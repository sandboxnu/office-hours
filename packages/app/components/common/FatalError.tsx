import { apm } from "@elastic/apm-rum";
import Result from "antd/lib/result";
import { ReactElement, useEffect } from "react";

export function FatalError({ error }: { error: Error }): ReactElement {
  useEffect(() => window && apm.captureError(error), [error]);
  return (
    <Result
      status="500"
      title="We hit an unexpected error."
      subTitle="Sorry about that! A report has automatically been filed. Try refreshing the page."
    />
  );
}
