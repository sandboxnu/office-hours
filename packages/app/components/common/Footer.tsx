import { ReactElement } from "react";
import styled from "styled-components";

const FullWidth = styled.footer`
  width: 100%;
  background: #ebebeb;

  flex-shrink: 0;
  padding: 12px 64px;

  display: flex;
  justify-content: space-between;
`;

export function Footer(): ReactElement {
  return (
    <FullWidth>
      <div>Made by Khoury College Edited by kevin</div>
    </FullWidth>
  );
}
