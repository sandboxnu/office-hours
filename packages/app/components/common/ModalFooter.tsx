import { Button, ButtonProps } from "antd";
import { ReactElement, ReactNode } from "react";
import styled from "styled-components";
import { useIsMobile } from "../../hooks/useIsMobile";

interface ModalFooterProps {
  onCancel?: () => void;
  onOk?: () => void;
  okText?: ReactNode;
  cancelText?: ReactNode;
  okButtonProps?: ButtonProps;
  cancelButtonProps?: ButtonProps;
}

const FooterContainer = styled.div`
  display: flex;
  justify-content: flex-end;
`;

const FooterButton = styled(Button)`
  @media (max-width: 650px) {
    flex-grow: 1;
  }
`;

// A mobile-friendly footer for modals
export default function ModalFooter({
  onCancel,
  onOk,
  okText = "OK",
  cancelText = "Cancel",
  okButtonProps = {},
  cancelButtonProps = {},
}: ModalFooterProps): ReactElement {
  const isMobile = useIsMobile();
  return (
    <FooterContainer>
      <FooterButton
        onClick={onCancel}
        size={isMobile ? "large" : "middle"}
        {...cancelButtonProps}
      >
        {cancelText}
      </FooterButton>
      <FooterButton
        type="primary"
        onClick={onOk}
        size={isMobile ? "large" : "middle"}
        {...okButtonProps}
      >
        {okText}
      </FooterButton>
    </FooterContainer>
  );
}
