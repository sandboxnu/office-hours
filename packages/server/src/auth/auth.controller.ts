import {
  Controller,
  Get,
  HttpStatus,
  Param,
  Query,
  Req,
  Res,
} from '@nestjs/common';
import { Request, Response } from 'express';
import { ConfigService } from '@nestjs/config';
import { AuthService } from './auth.service';
import { ERROR_MESSAGES } from '@koh/common';
import { JwtService } from '@nestjs/jwt';

@Controller('auth')
export class AuthController {
  private GOOGLE_AUTH_URL = 'https://accounts.google.com/o/oauth2/v2/auth';

  constructor(
    private jwtService: JwtService,
    private configService: ConfigService,
    private authService: AuthService,
  ) {}

  @Get(':method')
  auth(
    @Res() res: Response,
    @Param('method') auth_method: string,
  ): Response<{ redirectUri: string }> {
    switch (auth_method) {
      case 'google':
        res.cookie('organization.id', 1, {
          httpOnly: true,
          secure: this.isSecure(),
        });
        return res.status(200).send({
          redirectUri:
            `${this.GOOGLE_AUTH_URL}?client_id=${process.env.GOOGLE_CLIENT_ID}.apps.googleusercontent.com` +
            `&redirect_uri=${process.env.GOOGLE_REDIRECT_URI}&response_type=code&scope=openid%20profile%20email`,
        });
    }
  }

  @Get('callback/:method')
  async callback(
    @Res() res: Response,
    @Param('method') auth_method: string,
    @Query('code') auth_code: string,
    @Req() req: Request,
  ): Promise<Response<void>> {
    const organizationId = this.getCookie(req, 'organization.id');

    if (!organizationId) {
      res.redirect(`/auth/failed/40000`);
    } else {
      try {
        let payload: number;

        switch (auth_method) {
          case 'google':
            payload = await this.authService.loginWithGoogle(
              auth_code,
              Number(organizationId),
            );
            break;
          default:
            return res
              .status(HttpStatus.BAD_REQUEST)
              .send({ message: 'Invalid auth method' });
        }

        res.clearCookie('organization.id', {
          httpOnly: true,
          secure: this.isSecure(),
        });

        this.enter(res, payload);
      } catch (err) {
        res.redirect(HttpStatus.BAD_REQUEST, `/auth/failed/40001`);
      }
    }
  }

  private getCookie(req: Request, cookieName: string): string {
    const cookies: string[] = req.headers.cookie.split(';');
    for (const cookie of cookies) {
      const [name, value] = cookie.split('=');
      if (name.trim() === cookieName) {
        return value;
      }
    }
  }

  private async enter(res: Response, userId: number) {
    // Expires in 30 days
    const authToken = await this.jwtService.signAsync({
      userId,
      expiresIn: 60 * 60 * 24 * 30,
    });

    if (authToken === null || authToken === undefined) {
      return res
        .status(HttpStatus.INTERNAL_SERVER_ERROR)
        .send({ message: ERROR_MESSAGES.loginController.invalidTempJWTToken });
    }

    const isSecure = this.configService
      .get<string>('DOMAIN')
      .startsWith('https://');
    res
      .cookie('auth_token', authToken, { httpOnly: true, secure: isSecure })
      .redirect(HttpStatus.FOUND, `/courses`);
  }

  private isSecure(): boolean {
    return this.configService.get<string>('DOMAIN').startsWith('https://');
  }
}
