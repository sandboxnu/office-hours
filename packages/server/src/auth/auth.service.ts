import { AccountType } from '@koh/common';
import {
  BadRequestException,
  Injectable,
  InternalServerErrorException,
} from '@nestjs/common';
import { OAuth2Client } from 'google-auth-library';
import { OrganizationUserModel } from 'organization/organization-user.entity';
import { UserModel } from 'profile/user.entity';

@Injectable()
export class AuthService {
  client: OAuth2Client;

  constructor() {
    this.client = new OAuth2Client(
      process.env.GOOGLE_CLIENT_ID,
      process.env.GOOGLE_CLIENT_SECRET,
      process.env.GOOGLE_REDIRECT_URI,
    );
  }

  async loginWithGoogle(
    auth_code: string,
    organizationId: number,
  ): Promise<number> {
    try {
      const { tokens } = await this.client.getToken(auth_code);

      const ticket = await this.client.verifyIdToken({
        idToken: tokens.id_token,
        audience: `${process.env.GOOGLE_CLIENT_ID}.apps.googleusercontent.com`,
      });

      const payload = ticket.getPayload();

      if (!payload.email_verified) {
        throw new BadRequestException('Email not verified');
      }

      const user = await UserModel.findOne({ email: payload.email });

      if (user && user.password) {
        throw new BadRequestException(
          'User collisions with legacy account are not allowed',
        );
      }

      if (user) {
        return user.id;
      }

      if (!user) {
        const newUser = await UserModel.create({
          email: payload.email,
          firstName: payload.given_name,
          lastName: payload.family_name,
          photoURL: payload.picture,
          accountType: AccountType.GOOGLE,
        }).save();

        const userId = newUser.id;

        await OrganizationUserModel.create({
          organizationId,
          userId: userId,
        }).save();

        return userId;
      }

      throw new InternalServerErrorException('Unexpected error');
    } catch (err) {
      throw new BadRequestException(err.message);
    }
  }
}
