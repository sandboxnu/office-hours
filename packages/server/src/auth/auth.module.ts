import { Module } from '@nestjs/common';
import { AuthController } from './auth.controller';
import { AuthService } from './auth.service';
import { JwtStrategy } from 'login/jwt.strategy';
import { JwtModule } from '@nestjs/jwt';
import { ConfigModule, ConfigService } from '@nestjs/config';

@Module({
  imports: [
    JwtModule.registerAsync({
      imports: [ConfigModule],
      inject: [ConfigService],
      useFactory: async (configService: ConfigService) => ({
        secret: configService.get('JWT_SECRET'),
      }),
    }),
  ],
  controllers: [AuthController],
  providers: [AuthService, JwtStrategy],
})
export class AuthModule {}
