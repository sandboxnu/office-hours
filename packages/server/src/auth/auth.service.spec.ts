import { Test, TestingModule } from '@nestjs/testing';
import { AuthService } from './auth.service';
import { TestConfigModule, TestTypeOrmModule } from '../../test/util/testUtils';
import { UserModel } from 'profile/user.entity';
import { Connection } from 'typeorm';
import { OrganizationFactory } from '../../test/util/factories';
import { AccountType } from '@koh/common';

// Extend the OAuth2Client mock with additional methods
jest.mock('google-auth-library', () => {
  const actualLibrary = jest.requireActual('google-auth-library');

  class MockOAuth2Client extends actualLibrary.OAuth2Client {
    async getToken(code: string): Promise<any> {
      if (code === 'valid_code') {
        return Promise.resolve({ tokens: { id_token: 'valid_token' } });
      } else {
        return Promise.resolve({ tokens: { id_token: 'mocked_token' } });
      }
    }

    async verifyIdToken(options: any): Promise<any> {
      if (options.idToken !== 'valid_token') {
        return Promise.resolve({
          getPayload: () => ({
            email_verified: false,
            email: 'mocked_email@example.com',
            given_name: 'John',
            family_name: 'Doe',
            picture: 'mocked_picture_url',
          }),
        });
      } else {
        return Promise.resolve({
          getPayload: () => ({
            email_verified: true,
            email: 'mocked_email@example.com',
            given_name: 'John',
            family_name: 'Doe',
            picture: 'mocked_picture_url',
          }),
        });
      }
    }
  }

  return {
    OAuth2Client: MockOAuth2Client,
  };
});

describe('AuthService', () => {
  let service: AuthService;
  let conn: Connection;

  beforeAll(async () => {
    const module: TestingModule = await Test.createTestingModule({
      imports: [TestTypeOrmModule, TestConfigModule],
      providers: [AuthService],
    }).compile();

    service = module.get<AuthService>(AuthService);
    conn = module.get<Connection>(Connection);
  });

  afterAll(async () => {
    await conn.close();
  });

  beforeEach(async () => {
    await conn.synchronize(true);
  });

  describe('loginWithShibboleth', () => {
    it('should throw an error when user already exists with password', async () => {
      await UserModel.create({
        email: 'mocked_email@example.com',
        password: 'test_password',
      }).save();

      await expect(
        service.loginWithShibboleth(
          'mocked_email@example.com',
          'student@ubc.ca',
          'John',
          'Doe',
          1,
        ),
      ).rejects.toThrowError(
        'User collisions with legacy account are not allowed',
      );
    });

    it('should throw an error when user already exists with other account type', async () => {
      await UserModel.create({
        email: 'mocked_email@example.com',
        accountType: AccountType.GOOGLE,
      }).save();

      await expect(
        service.loginWithShibboleth(
          'mocked_email@example.com',
          'student@ubc.ca',
          'John',
          'Doe',
          1,
        ),
      ).rejects.toThrowError(
        'User collisions with other account types are not allowed',
      );
    });

    it('should return user id when user already exists without password', async () => {
      const user = await UserModel.create({
        email: 'mocked_email@example.com',
        accountType: AccountType.SHIBBOLETH,
      }).save();

      const userId = await service.loginWithShibboleth(
        'mocked_email@example.com',
        'student@ubc.ca',
        'John',
        'Doe',
        1,
      );
      expect(userId).toEqual(user.id);
    });

    it('should create a new user when user does not exist', async () => {
      const organization = await OrganizationFactory.create();

      const userId = await service.loginWithShibboleth(
        'mocked_email@example.com',
        'student@ubc.ca',
        'John',
        'Doe',
        organization.id,
      );
      const user = await UserModel.findOne(userId);
      expect(user).toMatchSnapshot();
    });
  });

  describe('loginWithGoogle', () => {
    it('should throw an error when email is not verified', async () => {
      await expect(service.loginWithGoogle('invalid_token', 1)).rejects.toThrow(
        'Email not verified',
      );
    });

    it('should throw an error when user already exists with password', async () => {
      await UserModel.create({
        email: 'mocked_email@example.com',
        password: 'test_password',
      }).save();

      await expect(
        service.loginWithGoogle('valid_code', 1),
      ).rejects.toThrowError(
        'User collisions with legacy account are not allowed',
      );
    });

    it('should throw an error when user already exists with other account type', async () => {
      await UserModel.create({
        email: 'mocked_email@example.com',
        accountType: AccountType.SHIBBOLETH,
      }).save();

      await expect(
        service.loginWithGoogle('valid_code', 1),
      ).rejects.toThrowError(
        'User collisions with other account types are not allowed',
      );
    });

    it('should return user id when user already exists without password', async () => {
      const user = await UserModel.create({
        email: 'mocked_email@example.com',
        accountType: AccountType.GOOGLE,
      }).save();

      const userId = await service.loginWithGoogle('valid_code', 1);
      expect(userId).toEqual(user.id);
    });

    it('should create a new user when user does not exist', async () => {
      const organization = await OrganizationFactory.create();

      const userId = await service.loginWithGoogle(
        'valid_code',
        organization.id,
      );
      const user = await UserModel.findOne(userId);
      expect(user).toMatchSnapshot();
    });
  });
});
