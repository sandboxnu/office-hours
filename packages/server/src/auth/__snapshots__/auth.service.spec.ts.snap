// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthService loginWithGoogle should create a new user when user does not exist 1`] = `
UserModel {
  "accountType": "google",
  "defaultMessage": null,
  "desktopNotifsEnabled": false,
  "email": "mocked_email@example.com",
  "firstName": "John",
  "hideInsights": null,
  "id": 1,
  "includeDefaultMessage": true,
  "insights": Array [
    "TotalStudents",
    "TotalQuestionsAsked",
    "MedianWaitTime",
    "QuestionTypeBreakdown",
    "MostActiveStudents",
    "QuestionToStudentRatio",
    "MedianHelpingTime",
  ],
  "lastName": "Doe",
  "name": "John Doe",
  "password": null,
  "phoneNotifsEnabled": false,
  "photoURL": "mocked_picture_url",
  "sid": null,
  "userRole": "user",
}
`;
